Improve synchronization of track metadata and file tags

uklotzde opened the pull request:
This PR introduces a simpler, more robust, and extensible synchronization between track metadata that is stored in the database and file tags. We currently only store a subset of properties in the database.

I have used and tested this code for many weeks and didn't notice any issues or lost file tags. It is also needed for synchronization of tracks with external libraries like #2282 that store an extended set of metadata.

## Reorganization

- Remove special case handling of certain properties from *TagLib* code
- Remove special case handling of certain properties from `TrackInfo`/`AlbumInfo`/`TrackMetadata`
- Move tag migration code from `TrackDAO` into `TrackRecord`
- Move all code for merging imported file tags into `TrackRecord`

## Import

`SoundSourceProxy::updateTrackFromSource()`: Import and merge file tags when loading track objects on the fly. This will populate all missing fields that are not (yet) stored in the database.

## Export

`Track::exportMetadata()`: Import and merge file tags once again before exporting metadata into file tags. We have to reload the file tags anyway to check for differences as we already do now, because files should only be modified if differences are detected. Additional debug logs show both the existing and the modified metadata before writing file tags.

uklotzde committed:
Import and merge missing track metadata from file tags

uklotzde commented:
I would like to get this into 2.3 to avoid maintaining 2 largely different versions.

ronso0 commented:
Could you share a simple routine to test this?

uklotzde commented:
Even if EXTRA_METADATA is enabled (a prerequisite) the extended properties are not visible in Mixxx. You should not notice any changes, unless you send data to aoide. The improvements are mainly behind the scenes and ensure that newly added fields in the Mixxx database get populated lazily from file tags as expected.

I will think about enabling some trace logs to report all properties when loading a track.

daschuer commented on src/track/trackrecord.cpp:
```diff
@@ -47,4 +47,85 @@ bool TrackRecord::updateGlobalKeyText(
+// yet. The added column "tracktotal" has been initialized with the
+// default value "//".
+// See also: Schema revision 26 in schema.xml
+const QString kReloadTrackTotal = "//";
```
QStringLiteral

daschuer replied:
QStringLiteral

daschuer commented on src/track/trackmetadatataglib.cpp:
```diff
@@ -66,6 +66,13 @@ bool checkID3v2HeaderVersionSupported(const TagLib::ID3v2::Header& header) {
+// as written by MusicBrainz Picard 2.1.3 although the mapping table
+// doesn't mention any "http://" prefix.
+// See also: https://picard.musicbrainz.org/docs/mappings
+const QString kMusicBrainzOwner = "http://musicbrainz.org";
```
constexpr QString kMusicBrainzOwner = QStringLiteral("http://musicbrainz.org");

daschuer commented on src/track/trackrecord.h:
```diff
@@ -87,6 +87,12 @@ class TrackRecord final {
 
+    // Merge the current metadata from the file with the subset of metadata
+    // that is stored in the library. This is required to delete any tags
+    // from the file that are not (yet) stored in the library database!
```
Do we really delete user data from the users file, because we do not support it?

uklotzde replied:
The comment is wrong, the word "NOT" got lost somehow :see_no_evil:

daschuer replied:
Ah, than it makes sense. :-)

daschuer commented on src/track/trackrecord.cpp:
```diff
@@ -47,4 +47,85 @@ bool TrackRecord::updateGlobalKeyText(
+    mergeNullableMetadataProperty(mergedAlbumInfo.refMusicBrainzReleaseId(), importedAlbumInfo.getMusicBrainzReleaseId());
+    mergeNullableMetadataProperty(mergedAlbumInfo.refRecordLabel(), importedAlbumInfo.getRecordLabel());
+#else
+    Q_UNUSED(mergedAlbumInfo);
```
Why can't we put the declaration also into the #if defined(__EXTRA_METADATA__) block?

uklotzde replied:
Of course. Maybe I have left it outside of the #ifdef to show that only the merge = conditional-copy operations for individual properties differ. In this (special) case the set of currently supported properties is empty.

daschuer replied:
if there is a reason, keep it like it is.

uklotzde replied:
Not really, already simplified.

daschuer commented on src/track/trackrecord.cpp:
```diff
@@ -47,4 +47,85 @@ bool TrackRecord::updateGlobalKeyText(
+}
+
+template<typename T>
+void mergeNullableMetadataProperty(
```
I think this is a plain conditional copy, not a merge. Merge implies that a part is kept. 
copyIfNotNull() or so?

daschuer reviewed (commented):
Some first comments. 
Unfortunately, the functional changes are hard to track.

uklotzde committed:
Use QStringLiteral

uklotzde committed:
Fix wrong comment

uklotzde committed:
Join conditional compilation regions

uklotzde committed:
Rename conditional copy operation

uklotzde committed:
Add missing merge operation for SeratoMarkers2

uklotzde committed:
Write "Serato Markers2" ID3v2 frame

uklotzde committed:
Merge branch 'master' of git@github.com:mixxxdj/mixxx.git into dev_library_metadata

uklotzde committed:
Merge branch 'master' of git@github.com:mixxxdj/mixxx.git into dev_library_metadata

Holzhaus commented:
Btw we should really add a flag to activate EXTRA_METADATA and enable it for the CI builds so that the code is built and tested on Travis and Appveyor.

uklotzde commented:
@Holzhaus We currently don't have a dedicated feature flag in master that enables the __EXTRA_METADATA__ code, only PR #2282 makes use of this. Any ideas?

uklotzde commented:
@daschuer I agree that the changes compared to the current state are hard to track. It is a major redesig.

Currently all the code for synchronizing metadata from the database with file tags is scattered all around. We have special case handling on different levels in TrackDAO and the low-level TagLib code, where it does not belong.

Holzhaus commented:
@uklotzde Have a look at this branch: https://github.com/Holzhaus/mixxx/tree/extra-metadata-flag

It's untested though, since I'm currently travelling and had to use my phone. Using vim on a touchscreen ain't easy btw ;-)

uklotzde commented:
@Holzhaus I hesitated to add a new feature flag without an actual use case apart from aoide. But we could do this.

Holzhaus commented:
The flag would obviously be disabled for release builds, but it would make it easier to test this. How do you test it currently? Do you put a `#define __EXTRA_METADATA__` somewhere? Or do you `export CXXFLAGS=__EXTRA_METADATA__`?

uklotzde committed:
Merge branch 'master' of git@github.com:mixxxdj/mixxx.git into dev_library_metadata

uklotzde committed:
Test merging of imported track metadata

- Preserve existing properties
- Replace missing extra properties

uklotzde committed:
Merge branch 'master' of git@github.com:mixxxdj/mixxx.git into dev_library_metadata

uklotzde committed:
Delete undefined member functions

uklotzde committed:
Merge branch 'master' of git@github.com:mixxxdj/mixxx.git into dev_library_metadata

uklotzde committed:
Initialize undefined coverart locations in database with NULL

...instead of with an empty string. Both work, but NULL is more explicit.

uklotzde committed:
Use CoverInfoRelative instead of CoverInfo

uklotzde committed:
Use CoverArtUtils::guessCoverInfo() when updating track from source

...to consider not only embedded cover art but also image files in
the folder of the track's file. Otherwise the source of the cover
info would be reset to UNKNOWN.

uklotzde commented:
I found some issues in the way cover art is guessed. Files in the folder were not considered if reloading metadata from source, only embedded covers. The CoverInfo *source* of all files without embedded cover is reset to UNKNOWN and the LibraryScanner will revisit them on the next run. Not a big deal but annoying.

Due to merge conflicts I will not fix it on master, but only in this branch.

daschuer commented on src/track/bpm.h:
```diff
@@ -25,6 +25,12 @@ class Bpm final {
 
     // Adjusts floating-point values to match their string representation
     // in file tags to account for rounding errors.
+    // NOTE(2019-02-19, uklotzde): The pre-normalization cannot prevent
```
Why can't we handle this case?
Do we actually repetitive write metadata if we detect 128.1 but 128 is stored?

uklotzde replied:
ID3v2 is restricted to integer precision.

uklotzde replied:
For other tag formats the internal 64-bit (binary) floating point value is stored as a (decimal) string. This will cause rounding errors that we need to compensate for. Otherwise the metadata is always detected as modified even if the file tags are effectively unchanged.

daschuer replied:
The critical point is: Is a 128.1 BMP track repeatedly written, because the decimal place is lost. 
I don't think this is the case, but the comment implies that.

uklotzde replied:
That is in fact true. I added a TL;DR to emphasize this point!

We can't do anything about it if we don't want to reduce the precision for the bpm comparison in general. This would be the only option.

uklotzde replied:
I have switched the corresponding *modified* check to integer precision comparison. As a minor drawback fractional bpm changes are not detected as modified. But we avoid modifying many files that actually don't need to be modified.

Just recently I saw complaints by Serato users, because Serato is writing library details like the *play count* into their marker tags! Every unmodified file you play needs to be copied during the next backup and this strategy also reduces the lifetime of your SSD.

daschuer replied:
Sounds reasonable. Thank  you.

daschuer commented on src/track/track.cpp:
```diff
@@ -137,11 +137,11 @@ void Track::setTrackMetadata(
-            m_record.setMetadata(std::move(trackMetadata));
-            // Don't use trackMetadata after move assignment!!
+            m_record.setMetadata(std::move(importedMetadata));
+            // Don't use importedMetadata after move assignment!!
```
Can't we delete this comment?

uklotzde replied:
Of course, it is redundant. But since the data of this argument is moved away in the middle of a large scope, this might serve as a reminder to future developers. Unfortunately C++ does not prevent you from doing unintended things as Rust would do.

uklotzde replied:
This code has just been moved around, nothing changed.

daschuer replied:
OK

uklotzde commented on src/library/coverartutils.cpp:
```diff
@@ -80,32 +80,29 @@ QImage CoverArtUtils::loadCover(const CoverInfo& info) {
+        image = std::move(embeddedCover);
+    }
     if (!image.isNull()) {
-        // TODO() here we my introduce a duplicate hash code
```
The comment is redundant. Every hash function has a risk of collisions, we don't need to repeat that at every occurrence. There are more places where the image hash is calculated without such a comment, e.g. in TrackDAO.

daschuer replied:
OK

uklotzde commented on src/sources/metadatasourcetaglib.cpp:
```diff
@@ -26,12 +26,12 @@ Logger kLogger("MetadataSourceTagLib");
 
 // Appended to the original file name of the temporary file used for writing
-const QString kSafelyWritableTempFileSuffix = "_temp";
+const QString kSafelyWritableTempFileSuffix = QStringLiteral("_temp");
```
Nope, QString has no constexpr constructor.

daschuer replied:
OK

uklotzde commented on src/sources/metadatasourcetaglib.cpp:
```diff
@@ -26,12 +26,12 @@ Logger kLogger("MetadataSourceTagLib");
 // file. Should not be longer than kSafelyWritableTempFileSuffix to avoid
 // potential failures caused by exceeded path length.
-const QString kSafelyWritableOrigFileSuffix = "_orig";
+const QString kSafelyWritableOrigFileSuffix = QStringLiteral("_orig");
```
Nope

daschuer replied:
Ok

uklotzde committed:
Cache possible covers from last folder

Removes redundant code out of TrackDAO that doesn't belong there!

uklotzde commented:
- Removed redundant code from TrackDAO that doesn't belong there
- Don't try to extract the embedded cover twice if it is not available when reloading metadata

uklotzde committed:
Restate re-export of ID3 tags for fractional bpm values

uklotzde committed:
Add different options for comparing bpm values

uklotzde committed:
Reduce precision for comparing bpm values when exporting metadata

uklotzde committed:
Merge branch 'master' of git@github.com:mixxxdj/mixxx.git into dev_library_metadata

uklotzde committed:
Merge branch 'master' into dev_library_metadata

daschuer commented on src/track/bpm.h:
```diff
@@ -25,6 +25,15 @@ class Bpm final {
+    // or not.
+    // TL;DR: If metadata export is enabled ID3 tags will be rewritten
+    // for all files with a fractional bpm values even if their metadata
+    // has not been modified.
```
I think this comment need to be updated.

uklotzde committed:
Update comment

daschuer commented:
Thank you LGTM. I have also done some brief testing, so we can risk to merge this 1k PR
Thank you for the work.

Holzhaus commented on src/track/trackrecord.cpp:
```diff
@@ -47,4 +49,88 @@ bool TrackRecord::updateGlobalKeyText(
 
+namespace {
+
+void mergeReplayGainMetadataProperty(
```
I'm getting a warning now:
```
/home/jan/Projects/mixxx/src/track/trackrecord.cpp:54:6: Warnung: »void mixxx::{anonymous}::mergeReplayGainMetadataProperty(mixxx::ReplayGain&, const mixxx::ReplayGain&)« definiert, aber nicht verwendet [-Wunused-function]
   54 | void mergeReplayGainMetadataProperty(
      |      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

uklotzde replied:
Oh, damn. An #ifdef is missing. Fix will follow.
